sdks/java/extensions/sketching/src/main/java/org/apache/beam/sdk/extensions/sketching/ApproximateDistinct.java

```diff
@@ -0,0 +1,533 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.extensions.sketching;
+
+import com.clearspring.analytics.stream.cardinality.CardinalityMergeException;
+import com.clearspring.analytics.stream.cardinality.HyperLogLogPlus;
+import com.google.auto.value.AutoValue;
+import com.google.common.base.MoreObjects;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import javax.annotation.Nullable;
+
+import org.apache.beam.sdk.coders.ByteArrayCoder;
+import org.apache.beam.sdk.coders.Coder;
+import org.apache.beam.sdk.coders.CoderException;
+import org.apache.beam.sdk.coders.CustomCoder;
+import org.apache.beam.sdk.coders.KvCoder;
+import org.apache.beam.sdk.transforms.Combine;
+import org.apache.beam.sdk.transforms.Combine.CombineFn;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.transforms.display.DisplayData;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.PCollection;
+
+/**
+ * {@link PTransform}s for computing the approximate number of distinct elements in a stream.
+ *
+ * <p>This class relies on the HyperLogLog algorithm, and more precisely HyperLogLog+,
+ * the improved version of Google.
+ *
+ * <p>The {@link CombineFn} is also exposed directly so it can be used as a state cell in a
+ * stateful {@link org.apache.beam.sdk.transforms.ParDo}. See {@link ApproximateDistinctFn}.
+ *
+ * <p>Three parameters can be tuned according to the processing context :
+ * <ul>
+ *    <li> Input Coder : The coder of the objects to combine.
+ *    <li> Precision : Controls the accuracy of the estimation. In general one can expect a
+ *    relative error of about {@code 1.1 / sqrt(2^p)}.
+ *    <li> Sparse Precision : Used to create a sparse representation in order to optimize memory
+ *    and improve accuracy at small cardinalities ({@code < 12000}).
+ * </ul>
+ *
+ * <h2>Using the Transforms</h2>
+ *
+ * <p>By default the Input Coder is inferred at runtime. One should specify a coder only if
+ * custom type is used.
+ * <br>By default the precision is set to {@code 12} for a relative error of around {@code 2%}.
+ * <br>By default the sparse representation is not used. One should use it if the number of
+ * distinct elements in the stream may be less than {@code 12000}.
+ *
+ * <h4>Example 1 : globally default use</h4>
+ * <pre>{@code
+ * PCollection<Integer> input = ...;
+ * PCollection<HyperLogLogPlus> hllSketch = input.apply(ApproximateDistinct.<Integer>globally());
```

Comment:
Hm, I wonder if it makes sense to make this transform's and CombineFn's result be a `Long` instead of a `HyperLogLogPlus` (and use HLLP just as the accumulator type), because really the only thing a user can do with the `HyperLogLogPlus` is call `.cardinality()` on it - am I missing something?

Reply:
Originally I wanted to return a Long but then I thought it could be useful for users to have the complete structure so they can for instance save it in a database or send it across another platform or whatever they can think of.

Reply:
I think you are right that there can be use cases for letting people access the HLLP directly. However, this would be a very advanced use case, to be used by people who care that they are dealing specifically with HLLP rather than simply computing approximate distinct counts.

How about this:

- Let's provide `ApproximateDistinct.hyperLogLogPlusCombineFn()` that is a `CombineFn` that returns a `HyperLogLogPlus`. This is what users will use for those advanced cases.
- Let's also provide a transform `ApproximateDistinct.globally()` that returns a `Long`, uses `hLLPCF()` under the hood, and make documentation focus on this transform, with `hyperLogLogPlusCombineFn()` mentioned as a footnote for advanced use cases.

Reply:
That is a quite complex but nice idea, I can think about that.